### DISCUSSION TITLE
Bug? Specific WAV files no longer play in Winamp 5.9 with libsndfile 1.1.0

### DISCUSSION THREAD

**Discussion by djegg** [libsndfile/libsndfile#872] [Q&A]
Hello

For [Winamp 5.9](http://forums.winamp.com/showthread.php?t=458120), we upgraded our Waveform Decoder (in_wave.dll) to use libsndfile v1.1.0

Previously, in Winamp 5.6.x to 5.8 we were using v1.0.25

This sample wav file (16bit signed stereo 44.1kHz CDQ / with RIFF header) played ok in 5.6/5.8, but doesn't play in 5.9
https://meggamusic.co.uk/winamp/test/WAV-06-Something_About_The_Way_You_Look_Tonight.zip

We have determined that the issue is metadata-parsing-related to an unknown chunk marker... 

![chunk](https://user-images.githubusercontent.com/6439427/189942393-a0117e3d-6a95-4816-9404-3f2554a268a3.png)

When calling the function ‘sf_wchar_open’, the returned pointer is null, causing the parser to exit.

Note that all other .wav files we've tested which played ok in previous versions still play ok in 5.9, it's just this specific format/criteria where there's a problem.

Is this something we can address at our end (in e.g. in_wave) or is it something you can fix in libsndfile which we can include for Winamp 5.9.1 or later?

Thanks for any help/feedback!

**Comment by evpobr** [comment]
Hi @djegg . 1.0.25 is was released more than 10 years ago. I'll try to find the latest version where everything worked, maybe this way we'll find a clue. But it will take time.

**Comment by evpobr** [comment]
Yep, it was right after 1.0.25 in 1.0.26:

```
Error : Not able to open input file .\WAV-06-Something About the Way You Look Tonight.wav.
File : .\WAV-06-Something About the Way You Look Tonight.wav
Length : 36710196
RIFF : 36710188
WAVE
LIST : 128
  INFO
    IPRD : Chinese Garden
    IART : Richard Clayderman & Nicolas De Angelis
    INAM : Something About the Way You Look Tonight
**** Bad chunk length 128 sbould be 126
  Unknown chunk marker at position 146. Resynching.
  Unknown chunk marker at position 151. Resynching.
*** Unknown chunk marker (20001) at position 156. Exiting parser.

Error in WAV file. No 'data' chunk marker.
```

**Comment by evpobr** [comment]
Well, the chance is small. 1.0.26. This version was developed for almost five years and the then developer no longer helps us.

**Comment by djegg** [comment]
Thank you for getting back to me.

Yes, however long it takes, if even possible at all.

There's only been 3 Winamp releases in the past ~10 years...
v5.6.66 was in 2013, 5.8 in 2018 (but no update for libsndfile), and 5.9 last week.

At first we thought the file was just corrupt, but it also plays ok in VLC, MPC-HC, Groove, Aimp, fb2k, WMP, etc.

Thanks again!

**Comment by evpobr** [comment]
Yes, i read about new WinAmp version yesterday and was very surprised.

**Comment by evpobr** [comment]
@SoapGentoo , @arthurt , maybe you have ideas how to help?

**Reply by SoapGentoo** [reply]
`git bisect`?

**Comment by arthurt** [comment]
My initial thought was that this WAV file was malformed, and that (judging by the logs) in v1.0.25 we used to be more tolerant in the RIFF chunk parser.

However, that actually appears not to be the case, that WAV file is actual well-formed, and we are parsing it wrong.

First 512 bytes of the file.
```
00000000  52 49 46 46 2c 27 30 02  57 41 56 45 4c 49 53 54  |RIFF,'0.WAVELIST|
00000010  80 00 00 00 49 4e 46 4f  49 50 52 44 0f 00 00 00  |....INFOIPRD....|
00000020  43 68 69 6e 65 73 65 20  47 61 72 64 65 6e 00 00  |Chinese Garden..|
00000030  49 41 52 54 28 00 00 00  52 69 63 68 61 72 64 20  |IART(...Richard |
00000040  43 6c 61 79 64 65 72 6d  61 6e 20 26 20 4e 69 63  |Clayderman & Nic|
00000050  6f 6c 61 73 20 44 65 20  41 6e 67 65 6c 69 73 00  |olas De Angelis.|
00000060  49 4e 41 4d 29 00 00 00  53 6f 6d 65 74 68 69 6e  |INAM)...Somethin|
00000070  67 20 41 62 6f 75 74 20  74 68 65 20 57 61 79 20  |g About the Way |
00000080  59 6f 75 20 4c 6f 6f 6b  20 54 6f 6e 69 67 68 74  |You Look Tonight|
00000090  00 00 00 00 66 6d 74 20  10 00 00 00 01 00 02 00  |....fmt ........|
000000a0  44 ac 00 00 10 b1 02 00  04 00 10 00 64 61 74 61  |D...........data|
000000b0  80 26 30 02 00 00 00 00  00 00 00 00 00 00 00 00  |.&0.............|
000000c0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
```
Starts with a RIFF header, then a LIST chunk, inside the list is an INFO list-identifier, then a bunch of info txt strings.

INFO chunks are a sequence of 4-byte ID, 4-byte length (LE), text of length, padded to a WORD boundary.

Isolating just the LIST chunk:
```
$ cat file.WAV | dd bs=1 skip=12 count=136 status=none | hexdump -C
00000000  4c 49 53 54 80 00 00 00  49 4e 46 4f 49 50 52 44  |LIST....INFOIPRD|
00000010  0f 00 00 00 43 68 69 6e  65 73 65 20 47 61 72 64  |....Chinese Gard|
00000020  65 6e 00 00 49 41 52 54  28 00 00 00 52 69 63 68  |en..IART(...Rich|
00000030  61 72 64 20 43 6c 61 79  64 65 72 6d 61 6e 20 26  |ard Clayderman &|
00000040  20 4e 69 63 6f 6c 61 73  20 44 65 20 41 6e 67 65  | Nicolas De Ange|
00000050  6c 69 73 00 49 4e 41 4d  29 00 00 00 53 6f 6d 65  |lis.INAM)...Some|
00000060  74 68 69 6e 67 20 41 62  6f 75 74 20 74 68 65 20  |thing About the |
00000070  57 61 79 20 59 6f 75 20  4c 6f 6f 6b 20 54 6f 6e  |Way You Look Ton|
00000080  69 67 68 74 00 00 00 00                           |ight....|
00000088              ^^ ^^ ^^^^^ 
                      |  |  |
                      |  |  Pad to DWORD boundary, end of LIST
                      |  Pad to WORD boundary, end of INAM
                      End of INAM text (null terminated string)

```
We know the chunk is 136 bytes long, as the length at byte offset 0x4 is 128, plus 8 for the four byte id and length.

Looking at the last string, INAM, it says it is 41 bytes long. List text strings by spect only start on even DOS WORD (aka 16-bit) boundaries. So the next text string identifier should begin after 42 bytes. That's at offset 0x86 of the chunk.

Can you see the problem? There are only two bytes left in the chunk. I'm guessing that the code that created this header padded the LIST chunk so that it ends on a 4-byte boundary, which is allowed. However, the list chunk parser, while it does check for how much of the chunk is left, assumes that it can always read the next identifier and length before checking that boundary. What ends up happening is at:

```
00000080  59 6f 75 20 4c 6f 6f 6b  20 54 6f 6e 69 67 68 74  |You Look Tonight|
00000090  00 00 00 00 66 6d 74 20  10 00 00 00 01 00 02 00  |....fmt ........|
000000a0  44 ac 00 00 10 b1 02 00  04 00 10 00 64 61 74 61  |D...........data|
```
bytes 0x92 - 0x96 [0x00, 0x00, 0x66, 0x6d] are read as the next text marker, and bytes 0x96 - 0x9a [0x74, 0x20, 0x10, 0x00] are read as it's length, which is 1056884. This is what causes the log message:
```
    ***  : 1056884
```

The code then detects that 1056884 is longer than the remaining LIST chunk, and stops parsing the LIST chunk and incorrectly rewinds to 0x98 instead of 0x94 like it should. Now the next chunk the parser sees is actually the length marker for the `fmt ` chunk [0x10, 0x00, 0x00, 0x00], causing

```
*** Unknown chunk marker (10) at position 152. Exiting parser.
```

And it barfs.

**Comment by arthurt** [comment]
So, this is definitely a bug. The function `wavlike_subchunk_parse ()` should be changed to

1) not try and read the next text marker and length if there are not enough bytes left in the chunk
2) correctly rewind in the case of a list member that is too long.

**Reply by arthurt** [reply]
Created issue #873

**Comment by evpobr** [comment]
Sorry I didn't answer for a long time, I got sick.

**Comment by evpobr** [comment]
An amazing investigation @arthurt !

**Comment by arthurt** [comment]
Related to #374

**Comment by evpobr** [comment]
@djegg, check latest master please.

**Reply by djegg** [reply]
Thanks! I will do.
Is it in vcpkg?

**Reply by evpobr** [reply]
No, master branch.

**Reply by evpobr** [reply]
@djegg 

[libsndfile-1.1.0-d8503a6-win32.zip](https://github.com/libsndfile/libsndfile/files/10047015/libsndfile-1.1.0-d8503a6-win32.zip)
[libsndfile-1.1.0-d8503a6-win64.zip](https://github.com/libsndfile/libsndfile/files/10047016/libsndfile-1.1.0-d8503a6-win64.zip)

**Reply by djegg** [reply]
Sorry for the delay.
I can confirm the fix.
We're using vcpkg for our external libs, so we'll need to implement the fix via a patch.
Thanks again!
